corehq/apps/locations/models.py

```diff
@@ -468,6 +468,39 @@ def get_descendants(self, include_self=False, **kwargs):
             where, **kwargs
         )
 
+    @staticmethod
+    def get_bulk_ancestors(domain, location_ids, **kwargs):
```

Comment:
This doesn't seem like it belongs in the `SQLLocation` class - could you make it a standalone function near the report where it's used?

Reply:
hey, It could be a separate talk about to what extent we can put a method in class as far as the class methods are handling only one thing. But I have put this method here because this deals just with Location. and returns only the location info same as already existing `get_ancestors` method below just in bulk form.
But other face of this is, this new method is just used in only one report so considering that it might make sense to move it there. 
Whats your thought on this.

Reply:
Yeah, that's a good question, maybe we could bring it up at the next book club meeting to see what people think.  I don't think there's a hard and fast rule about it, and I think there's variation across different parts of the codebase.  Here's some of what I think about with Locations in particular:

Following the Django convention, methods which return individual locations or querysets of locations go on `SQLLocation.objects`.  Highly reusable filtering abstractions can also go on the model manager or queryset (check out `LocationQueriesMixin` and `LocationManager`).  This lets us create abstractions around concepts that should be consistent all across the website, such as what it means for a location to be accessible to a user, or how exactly searching for a location works.  I wrote a [blog post](https://www.dimagi.com/blog/querysets-for-business-logic-in-django-orm/) about the topic, actually.  These bits of logic can then be combined with other queryset methods as needed for the task at hand.

Our dbaccessors pattern has a combinatorial problem.  You often need slight variations on a theme, and implementing each as a standalone function means you get a lot of functions, for instance, getting apps with variation on returning the wrapped application, the unwrapped application, or the ids only gives you three functions, and if you want the same but for only app builds, you get six fuctions: `get_apps`, `get_app_ids`, `get_app_docs`, `get_app_builds`, `get_app_build_ids`, `get_app_build_docs`.  Django querysets avoid this problem by letting you chain the logic together, like `get_apps().builds().docs()`.  For that reason, I also prefer to avoid methods which combine query components together that aren't logically connected (`get_locations_and_children_ids` and `accessible_location_ids` violate this principle, since you can do `accessible_to_user().location_ids()`).

It's also nice to avoid using IDs directly, as that means whatever you're doing is probably not very performant.  Since we don't have users in SQL, we do end up needing to this a lot, though.  Still, I have seen places where we combine multiple bits of location queryset logic based on IDs, which means multiple queries, when instead we could've done everything with querysets.

If an operation performs a location query and it doesn't need to be combined with other queries, I think it's better as a standalone function.  They're a little easier to reason about, and to test and if necessary, replace.

I want to keep the `SQLLocation` class relatively small and manageable.  It's definitely bigger than it should be.  I think all methods of the class should deal with interacting with that particular row in the database, with rare exceptions like `to_json()`, `display_name` and other core aspects of an individual loction.  `case_sharing_group_object` is a violation of this, for example.

Reply:
Wow that's a long PR comment.

Reply:
Hey, That makes sense what you said. Specially the part which says different functions end up being created.
I am convinced to move this function back to the report to maintain the standard for SQLLocation